ttnn/cpp/ttnn/operations/experimental/reshape/reshape.cpp

```diff
@@ -0,0 +1,163 @@
+            shard_spec.shape[0] = new_shape[-1] > shard_shape[1] ? shard_shape[0] / mul_div : shard_shape[0] * mul_div;
+            shard_spec.shape[1] = new_shape[-1];
+
+            shard_spec_buffer.page_shape = {1, new_shape[-1]};
+            shard_spec_buffer.tensor2d_shape = {shard_spec.shape[0], 1};
+            shard_spec_buffer.set_shard_spec(shard_spec);
+
+            device_buffer->set_shard_spec(shard_spec_buffer);
+            device_storage.insert_buffer(device_buffer);
+
+            return Tensor(device_storage, new_shape, input_tensor.get_dtype(), input_tensor.get_layout(), std::nullopt);
+        }
+    } else {
+        const auto tile = input_tensor.get_tensor_spec().tile();
+        return Tensor(input_tensor.get_storage(), new_shape, input_tensor.get_dtype(), input_tensor.get_layout(), tile);
+    }
+}
+
+ttnn::Tensor tensor_reshape(const ttnn::Tensor& input_tensor, const ttnn::Shape& new_shape) {
+    ZoneScoped;
+    GraphTracker::instance().track_function_start("ttnn::experimental::unsafe_view", input_tensor, new_shape);
+    const auto& new_padded_shape = new_shape.padded_shape();
+
+    TT_ASSERT(
+        input_tensor.volume() == new_padded_shape.volume(),
+        "{} != {}",
+        input_tensor.volume(),
+        new_padded_shape.volume());
+    if (input_tensor.get_layout() == Layout::TILE) {
+        const auto tile = input_tensor.get_tensor_spec().tile();
+        TT_ASSERT(
+            new_padded_shape[-2] % tile.get_tile_shape()[0] == 0 &&
+            new_padded_shape[-1] % tile.get_tile_shape()[1] == 0 &&
+            "Expected a multiple of 32 for H, W (or -1 evaluating to such) in ttnn::experimental::unsafe_view()!");
+    }
+    auto output = std::visit(
+        [&input_tensor, &new_shape](auto&& storage) -> Tensor {
+            using T = std::decay_t<decltype(storage)>;
+            const auto& tensor = input_tensor;
+            if constexpr (std::is_same_v<T, MultiDeviceHostStorage>) {
+                return MultiDeviceHostStorage_visit(storage, input_tensor, new_shape);
+            }
+            if constexpr (std::is_same_v<T, MultiDeviceStorage>) {
+                return MultiDeviceStorage_visit(storage, input_tensor, new_shape);
+            }
+            if constexpr (std::is_same_v<T, DeviceStorage>) {
+                return DeviceStorage_visit(storage, input_tensor, new_shape);
+            } else {
+                if (input_tensor.get_layout() == Layout::ROW_MAJOR) {
+                    return Tensor(
+                        tensor.get_storage(), new_shape, tensor.get_dtype(), tensor.get_layout(), std::nullopt);
+                } else {
+                    const auto tile = input_tensor.get_tensor_spec().tile();
+                    return Tensor(tensor.get_storage(), new_shape, tensor.get_dtype(), tensor.get_layout(), tile);
+                }
+            }
+        },
```

Comment:
Any reason you refactored the paths here into separate functions? It's not meant to be reused somewhere else at all. Also, these function names are inconsistent in styling.

Reply:
Based on other reviewers' comments, I refactored the code here for the function to be cleaner and more readable. I have also fixed the styling.